onnxruntime/core/providers/cpu/math/einsum_utils/einsum_typed_compute_processor.cc

```diff
@@ -173,10 +194,16 @@ std::unique_ptr<Tensor> EinsumTypedComputeProcessor<T>::PairwiseOperandProcess(c
   left_permutation.insert(left_permutation.end(), ro.begin(), ro.end());
   if (EinsumOp::IsTransposeRequired(current_left ? current_left->Shape().GetDims().size() : left_dims.size(),
                                     left_permutation)) {
-    current_left = EinsumOp::Transpose(current_left ? *current_left : left,
-                                       current_left ? current_left->Shape().GetDims() : left_dims,
-                                       left_permutation, allocator_, einsum_ep_assets_,
-                                       device_transpose_func_);
+    if (current_left && IsTransposeReshapeForEinsum(left_permutation,
+                                                    current_left->Shape().GetDims(),
+                                                    reshaped_dims)) {
+      current_left->Reshape(reshaped_dims);
```

Comment:
nit: a comment stating why this is safe to do this might be useful to the code reader - This can be done because curent_* tensors (if they exist) are intermediate tensors and cannot be input tensors to the `Einsum` node itself (which are immutable). 

As a (useful or less useful) extension, I think we may be able to apply this optimization to `left` (and not `right`) because `left` may be an intermediate tensor as it may be an output of a previous pair-wise operation. But, it does need some "stateful" tracking and not sure if there is a case that warrants that at this moment.

Reply:
We should be able to apply it in all case but it is not easy with the current API to create another tensor sharing the same buffer but with a different shape. That would be easier to write. We could also get rid of the argument *input_shape_override* when calling transpose.

Reply:
> Are we sure we one of the existing test cases cover this or is adding a test for this change safe ?

One case is not covered. I need to find a configuration where it applies.

Reply:
The final case is now covered by a unit test.